src/test/java/com/microsoft/sqlserver/jdbc/AlwaysEncrypted/MultiUserAKVTest.java

```diff
@@ -0,0 +1,643 @@
+        }
+        
+        SQLServerConnection.unregisterColumnEncryptionKeyStoreProviders();
+        Map<String, SQLServerColumnEncryptionKeyStoreProvider> providerMap = new HashMap<String, SQLServerColumnEncryptionKeyStoreProvider>();
+        providerMap.put(Constants.AZURE_KEY_VAULT_NAME, akvProvider);
+        SQLServerConnection.registerColumnEncryptionKeyStoreProviders(providerMap);
+
+        int customerId = 10;
+        String customerName = "Microsoft";
+        createTableForCustomProvider(AETestConnectionString, customProviderTableName, cekAkv);
+        insertData(customProviderTableName, customerId, customerName);        
+        
+        String sql = "SELECT CustomerId, CustomerName FROM " + customProviderTableName + " WHERE CustomerId = ?";
+
+        try (SQLServerConnection con = PrepUtil.getConnection(AETestConnectionString, AEInfo);
+                SQLServerPreparedStatement pstmt = (SQLServerPreparedStatement) TestUtils.getPreparedStmt(con, sql,
+                    SQLServerStatementColumnEncryptionSetting.Enabled)) {            
+            
+            pstmt.setInt(1, customerId);
+            
+            ResultSet rs = pstmt.executeQuery();
+            while (rs.next()) {
+                int intValue = rs.getInt(1);
+                String strValue = rs.getString(2);
+                assertTrue((customerId == intValue) && strValue.equalsIgnoreCase(customerName));
+            }
+            
+            // Clean up global custom providers
+            SQLServerConnection.unregisterColumnEncryptionKeyStoreProviders();
+            
+            // Register key store provider on statement level
+            providerMap.put(Constants.AZURE_KEY_VAULT_NAME, provider);
+            pstmt.registerColumnEncryptionKeyStoreProvidersOnStatement(providerMap);
+            
+            // Execute a query using provider from statement-level cache. this will cache the cek in the local cek cache
+            ResultSet rs2 = pstmt.executeQuery();
+            while (rs2.next()) {
+                int intValue = rs2.getInt(1);
+                String strValue = rs2.getString(2);
+                assertTrue((customerId == intValue) && strValue.equalsIgnoreCase(customerName));
+            }
+            
+            // Register invalid key store provider on statement level. This will overwrite the previous one.
+            SQLServerColumnEncryptionAzureKeyVaultProvider providerWithBadCred = new SQLServerColumnEncryptionAzureKeyVaultProvider("badApplicationID", "badApplicationKey");
+            providerMap.put(Constants.AZURE_KEY_VAULT_NAME, providerWithBadCred);
+            pstmt.registerColumnEncryptionKeyStoreProvidersOnStatement(providerMap);
+
+            // The following query should fail due to an empty cek cache and invalid credentials
+            try (ResultSet rs3 = pstmt.executeQuery()) {
+                int numberOfColumns = rs3.getMetaData().getColumnCount(); 

```

Comment:
unused variable warning in eclipse (also line 349, 363, 378, 391, 437, 446, 456, 466, 479

Reply:
removed